ReactNativeClient/lib/file-api-driver-amazon-s3.js

```diff
@@ -0,0 +1,361 @@
+const { basicDelta } = require('lib/file-api');
+const { basename } = require('lib/path-utils');
+const { shim } = require('lib/shim');
+const JoplinError = require('lib/JoplinError');
+
+const S3_MAX_DELETES = 1000;
+
+class FileApiDriverAmazonS3 {
+	constructor(api, s3_bucket) {
+		this.s3_bucket_ = s3_bucket;
+		this.api_ = api;
+	}
+
+	api() {
+		return this.api_;
+	}
+
+	requestRepeatCount() {
+		return 3;
+	}
+
+	makePath_(path) {
+		if (!path) return '';
+		return path;
+	}
+
+	hasErrorCode_(error, errorCode) {
+		if (!error || typeof error.code !== 'string') return false;
+		return error.code.indexOf(errorCode) >= 0;
+	}
+
+	// Need to make a custom promise, built-in promise is broken: https://github.com/aws/aws-sdk-js/issues/1436
+	async s3GetObject(key) {
+		return new Promise((resolve, reject) => {
+			this.api().getObject({
+				Bucket: this.s3_bucket_,
+				Key: key,
+			}, (err, response) => {
+				if (err) reject(err);
+				else resolve(response);
+			});
+		});
+	}
+
+	async s3ListObjects(key, cursor) {
+		return new Promise((resolve, reject) => {
+			this.api().listObjectsV2({
+				Bucket: this.s3_bucket_,
+				Prefix: key,
+				Delimiter: '/',
+				ContinuationToken: cursor,
+			}, (err, response) => {
+				if (err) reject(err);
+				else resolve(response);
+			});
+		});
+	}
+
+	async s3HeadObject(key) {
+		return new Promise((resolve, reject) => {
+			this.api().headObject({
+				Bucket: this.s3_bucket_,
+				Key: key,
+			}, (err, response) => {
+				if (err) reject(err);
+				else resolve(response);
+			});
+		});
+	}
+
+	async s3PutObject(key, body) {
+		return new Promise((resolve, reject) => {
+			this.api().putObject({
+				Bucket: this.s3_bucket_,
+				Key: key,
+				Body: body,
+			}, (err, response) => {
+				if (err) reject(err);
+				else resolve(response);
+			});
+		});
+	}
+
+	async s3UploadFileFrom(path, key) {
+		if (!shim.fsDriver().exists(path)) throw new Error('s3UploadFileFrom: file does not exist');
+		const body = await shim.fsDriver().readFile(path, 'Buffer');
```

Comment:
This is for `put` with `options.source === 'file'`, not sure if this will break on ReactNative/mobile apps.